onnx/src/ops/array/range.rs

```diff
@@ -0,0 +1,104 @@
+
+use crate::model::ParsingContext;
+use crate::pb::*;
+
+#[derive(Debug, Clone, new, Hash)]
+pub struct Range {
+    dtype: DatumType,
+}
+
+impl_dyn_hash!(Range);
+
+pub fn range(_ctx: &ParsingContext, _node: &NodeProto) -> TractResult<(Box<dyn InferenceOp>, Vec<String>)> {
+    Ok((Box::new(Range::new(DatumType::I64)), vec![])) // Model works on long (i64) elements
+}
+
+impl Range {
+    fn eval_t<T>(&self, mut inputs: TVec<Arc<Tensor>>) -> TractResult<TVec<Arc<Tensor>>>
+    where
+        T: Datum
+            + AsPrimitive<usize>
+            + Add<T, Output = T>
+            + Div<T, Output = T>
+            + Mul<T, Output = T>
+            + Sub<T, Output = T>,
+        usize: AsPrimitive<T>,
+    {
+        let (start, limit, delta) = args_3!(inputs);
+        let start = *start.to_scalar::<T>()?;
+        let limit = *limit.to_scalar::<T>()?;
+        let delta = *delta.to_scalar::<T>()?;
+        let value =
+            Array1::from_shape_fn(((limit - start) / delta).as_(), |ix| ix.as_() * delta + start);
+        Ok(tvec![value.into_arc_tensor()])
+    }
+}
+
+impl Op for Range {
+    fn name(&self) -> Cow<str> {
+        "Range".into()
+    }
+
+    op_onnx!();
+    not_a_typed_op!();
+}
+
+impl EvalOp for Range {
+    fn is_stateless(&self) -> bool {
+        true
+    }
+
+    fn eval(&self, inputs: TVec<Arc<Tensor>>) -> TractResult<TVec<Arc<Tensor>>> {
+        dispatch_numbers!(Self::eval_t(self.dtype)(self, inputs))
+    }
+}
+
+impl InferenceRulesOp for Range {
+    fn rules<'r, 'p: 'r, 's: 'r>(
+        &'s self,
+        s: &mut Solver<'r>,
+        inputs: &'p [TensorProxy],
+        outputs: &'p [TensorProxy],
+    ) -> InferenceResult {
+        check_input_arity(&inputs, 3)?;
+        check_output_arity(&outputs, 1)?;
+        s.equals(&inputs[0].datum_type, self.dtype)?;
+        s.equals(&inputs[1].datum_type, self.dtype)?;
+        s.equals(&inputs[2].datum_type, self.dtype)?;
+        s.equals(&outputs[0].datum_type, self.dtype)?;
+        s.equals(&inputs[0].rank, 0)?;
+        s.equals(&inputs[1].rank, 0)?;
+        s.equals(&inputs[2].rank, 0)?;
+        s.equals(&outputs[0].rank, 1)?;
+        Ok(())
+    }
+
+    as_op!();
+
+    fn to_typed(
+        &self,
+        _source: &InferenceModel,
+        node: &InferenceNode,
+        target: &mut TypedModel,
+        mapping: &HashMap<OutletId, OutletId>,
+    ) -> TractResult<TVec<OutletId>> {
+        if let (Some(start), Some(limit), Some(delta)) = (
+            target.outlet_fact(mapping[&node.inputs[0]])?.konst.as_ref(),
+            target.outlet_fact(mapping[&node.inputs[1]])?.konst.as_ref(),
+            target.outlet_fact(mapping[&node.inputs[2]])?.konst.as_ref(),
+        ) {
+            let mut value = dispatch_numbers!(Self::eval_t(start.datum_type())(
+                self,
+                tvec!(start.clone(), limit.clone(), delta.clone())
+            ))?;
+            Ok(tvec!(target.add_const(&*node.name, value.remove(0))?))
```

Comment:
I'm not sure this is the right call here. I would expect one of the inputs of the Range op to be the a symbolic dimension (as it will likely be the input length)... Making it accept symbolic dimensions is a bit of work though : as it will no longer be a constant, it will have to be moved to core or an -opl crate.

The impact on performance could be significant, it should allow you to run into_optimized() before fixing the input length (which is usefull for NLP as input as utterances tend to have varrying sizes).